File: jun23/BinarySearchTree.py
from Queue import Queue
class Node:
    def __init__(self , data):
        self.data = data
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self , root , x):
        if root is None:
            return Node(x)

        if x < root.data:
            root.left = self.insert(root.left , x)
        elif x > root.data:
            root.right = self.insert(root.right , x)
        else:
            print("Same Element Given")

        return root

    def delete(self, root, x):
        if root is None:
            return root

        if x < root.data:
            root.left = self.delete(root.left, x)

        elif x > root.data:
            root.right = self.delete(root.right, x)

        else:
            if root.left is None and root.right is None:
                return None

            elif root.left is None:
                return root.right

            elif root.right is None:
                return root.left

            else:
                s = self.getSuccessor(root.right)
                root.data = s.data
                root.right = self.delete(root.right, s.data)

        return root


    def getSuccessor(self, curr):
        while curr.left is not None:
            curr = curr.left
        return curr


    def inorder(self, root):
        if root is not None:
            self.inorder(root.left)
            print(root.data, end=" ")
            self.inorder(root.right)


    def levelorder(self, root):
        if root is None:
            return

        q = Queue()
        q.enqueue(root)

        while not q.isEmpty():
            curr = q.dequeue()

            print(curr.data, end=" ")

            if curr.left:
                q.enqueue(curr.left)

            if curr.right:
                q.enqueue(curr.right)

File: jul03/BinarySearchTree.java
package jul03;

import java.util.LinkedList;
import java.util.Queue;

class Node {
    int data;
    Node left;
    Node right;

    Node(int data) {
        this.data = data;
    }
}

public class BinarySearchTree {
    public static Node insert(Node root, int data) {
        if (root == null) {
            return new Node(data);
        }
        if (root.data > data) {
            root.left = insert(root.left, data);
        } else {
            root.right = insert(root.right, data);
        }
        return root;
    }

    public static void inorder(Node root){
        if (root == null) {
            return;
        }
        inorder(root.left);
        System.out.print(root.data + " ");
        inorder(root.right);
    }
    public static void levelOrder(Node root){
        if (root == null) {
            return;
        }
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
           Node curr = q.poll();
           System.out.print(curr.data+" ");
           if(curr.left != null){
            q.add(curr.left);
           }
           if(curr.right != null){
            q.add(curr.right);
           }
           

        }
    }
    public static void main(String[] args) {
        int[] data = {5 , 0 , 1 , 4 , 40 , 32 ,  39 , 41};
        Node root = new Node(data[0]);
        for (int i = 1; i < data.length; i++){
            insert(root, data[i]);
        }
        inorder(root);
        System.out.print("\n");
        levelOrder(root);
    }
}
